.github/workflows/ci.yml

```diff
@@ -0,0 +1,78 @@
+name: Continuous Integration
+
+on: [pull_request, workflow_dispatch]
+
+defaults:
+  run:
+    shell: bash
+
+concurrency:
+  group: ${{ github.workflow }}-${{ github.head_ref || github.ref }}
+  cancel-in-progress: true
+
+jobs:
+  pre-commit:
+    name: ruff and hooks.
+    runs-on: ubuntu-latest
+    steps:
+      - uses: actions/checkout@v4
+      - uses: actions/setup-python@v5
+        with:
+          python-version: '3.8'
+      - uses: pre-commit/action@v3.0.1
+        with:
+          extra_args: --all-files --show-diff-on-failure
+
+  ci:
+    name: CI
+    runs-on: ubuntu-latest
+    if: |
+      github.event_name == 'push' ||
+      (github.event_name == 'pull_request' && github.event.pull_request.draft == false)
+    strategy:
+      fail-fast: false
+      matrix:
+        python-version: ['3.8']
+        torch-version: ['1.11', '2.2']
+
+    steps:
+    - name: Checkout
+      uses: actions/checkout@v4
+      with:
+        fetch-depth: 0
+        lfs: false
+
+    - name: Set up Python
+      uses: actions/setup-python@v5
+      with:
+        python-version: ${{ matrix.python-version }}
+
+    - name: Cache dependency
+      id: cache-dependencies
+      uses: actions/cache@v4
+      with:
+        path: ~/.cache/pip
```

Comment:
This doesn't necessarily has something to do with this PR, we can also open an issue for it: currently runs with torch 2.2 take one minute less than with 1.11 because we've cached the dependencies for 2.2. I think which version is cached is random, probably based on which job was finished first the first time we ran it.

The easy fix would be to cache the different sets of dependencies with something similar to what we had in the PR introducing caching: https://github.com/sbi-dev/sbi/pull/1028#discussion_r1530476912. `${{ runner.os }}-pip-${{ matrix.python-version }}-${{ matrix.torch-version }}$`

Reply:
Nice catch! Let's also include this for this PR, since we are changing things anyway. I will have a look at the comment you linked in the next days.
To avoid misunderstandings: we want a separate cache for every realization of the test matrix, right? Is this actually still helpful/necessary when things have been moved to docker containers (there is another PR on that I think)? Anyways, if you already have an idea, feel free to edit the PR or post it as a comment :+1:

Reply:
Nice catch! I could integrate that in this PR though. So the line goes in as `key` I assume? The caching action just looks at whatever the key evaluates to and pastes in the matching deps I assume?

Reply:
Thanks for already doing it. I just added that we also look for keys which should at least contain some of the right dependencies.